url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/Index.java

```diff
@@ -0,0 +1,352 @@
+		for (int cChild = 0; cChild < m_children; cChild++) {
+			OverlapEntry e = new OverlapEntry();
+
+			e.m_id = cChild;
+			e.m_original = m_pMBR[cChild];
+			e.m_combined = m_pMBR[cChild].combinedRegion(r);
+			e.m_oa = e.m_original.getArea();
+			e.m_ca = e.m_combined.getArea();
+			e.m_enlargement = e.m_ca - e.m_oa;
+			entries[cChild] = e;
+
+			if (e.m_enlargement < me) {
+				me = e.m_enlargement;
+				best = cChild;
+			} else if (e.m_enlargement == me && e.m_oa < entries[best].m_oa) {
+				best = cChild;
+			}
+		}
+
+		if (me < SpatialIndex.EPSILON || me > SpatialIndex.EPSILON) {
+			int cIterations;
+
+			if (m_children > m_pTree.m_nearMinimumOverlapFactor) {
+				// sort entries in increasing order of enlargement.
+				Arrays.sort(entries, new OverlapEntryComparator());
+				cIterations = m_pTree.m_nearMinimumOverlapFactor;
+			} else {
+				cIterations = m_children;
+			}
+
+			// calculate overlap of most important original entries (near minimum overlap cost).
+			for (int cIndex = 0; cIndex < cIterations; cIndex++) {
+				double dif = 0.0;
+				OverlapEntry e = entries[cIndex];
+
+				for (int cChild = 0; cChild < m_children; cChild++) {
+					if (e.m_id != cChild) {
+						double f = e.m_combined.getIntersectingArea(m_pMBR[cChild]);
+						if (f != 0.0)
+							dif += f - e.m_original.getIntersectingArea(m_pMBR[cChild]);
+					}
+				} // for (cChild)
+
+				if (dif < leastOverlap) {
+					leastOverlap = dif;
+					best = cIndex;
+				} else if (dif == leastOverlap) {
+					if (e.m_enlargement == entries[best].m_enlargement) {
+						// keep the one with least area.
+						if (e.m_original.getArea() < entries[best].m_original.getArea())
+							best = cIndex;
+					} else {
+						// keep the one with least enlargement.
+						if (e.m_enlargement < entries[best].m_enlargement)
+							best = cIndex;
+					}
+				}
+			} // for (cIndex)
+		}
+
+		return entries[best].m_id;
+	}
+
+	protected void adjustTree(Node n, Stack<Integer> pathBuffer) {
+		m_pTree.m_stats.m_adjustments++;
+
+		// find entry pointing to old node;
+		int child;
+		for (child = 0; child < m_children; child++) {
+			if (m_pIdentifier[child] == n.m_identifier)
+				break;
+		}
+
+		// MBR needs recalculation if either:
+		//   1. the NEW child MBR is not contained.
+		//   2. the OLD child MBR is touching.
+		boolean b = m_nodeMBR.contains(n.m_nodeMBR);
+		boolean recalc = (!b) ? true : m_nodeMBR.touches(m_pMBR[child]);
+
+		m_pMBR[child] = (Region) n.m_nodeMBR.clone();
+
+		if (recalc) {
+			for (int cDim = 0; cDim < m_pTree.m_dimension; cDim++) {
+				m_nodeMBR.m_pLow[cDim] = Double.POSITIVE_INFINITY;
+				m_nodeMBR.m_pHigh[cDim] = Double.NEGATIVE_INFINITY;
+
+				for (int cChild = 0; cChild < m_children; cChild++) {
+					m_nodeMBR.m_pLow[cDim] = Math.min(m_nodeMBR.m_pLow[cDim], m_pMBR[cChild].m_pLow[cDim]);
+					m_nodeMBR.m_pHigh[cDim] = Math.max(m_nodeMBR.m_pHigh[cDim], m_pMBR[cChild].m_pHigh[cDim]);
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 257:267, 296:306.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.